### ISSUE TITLE
Revise error handling to make it easier for clients to distinguish function outcomes

### ISSUE THREAD

**Issue by hodgesrm** [vmware/go-vcloud-director#69]
Error handling in go-vcloud-director does not clearly distinguish clearly between different types of errors:
* Not finding something
* Failing due to a client-side error like bad host name
* Failing due to a problem on the server, e.g., not authorized, bad input, or a server failure

These are distinct outputs that require different handling; the first output may not even be an error.  In our current model clients must check strings to determine exactly what happened.  Here’s an example: finding Orgs by name.  In the current implementation two things can happen: 
* You find an Org and return it.
* Anything else is an error whose exact nature is denoted by the string value of the error message.
The following code from system.go illustrates:
```
func GetOrgByName(c *VCDClient, orgname string) (Org, error) {
        OrgHREF, err := getOrgHREF(c, orgname)
        if err != nil {
                return Org{}, fmt.Errorf("Cannot find the url of the org: %s", err)
        }
        u, err := url.ParseRequestURI(OrgHREF)
        if err != nil {
                return Org{}, fmt.Errorf("Error parsing org href: %v", err)
        }
        req := c.Client.NewRequest(map[string]string{}, "GET", *u, nil)
        resp, err := checkResp(c.Client.Http.Do(req))
        if err != nil {
                return Org{}, fmt.Errorf("error retreiving org: %s", err)
        }

        o := NewOrg(&c.Client)
        if err = decodeBody(resp, o.Org); err != nil {
                return Org{}, fmt.Errorf("error decoding org response: %s", err)
        }
        return *o, nil
}
```
It's hard to write a client that uses this call to see if an Org already exists, because you can't easily tell the difference between the call failing and the fact that the Org is not there.  This is because you have to parse a string that could change and that you might not have right in the first place. 

We need to look carefully at our function outputs.  In general a function like the preceding example should do one of the following:
* If you find the org return it.
* If the org is not there return nil.
* If you return an error return a _typed_ value that corresponds (if possible) to the HTTP status.  HTTP status codes are well-understood and easily extractable from any error.  We can add a few client side codes as well.  But the point is that the error should be a type that the compiler can confirm so that you have proof your client is looking for something that is actually there. 

In short, we should think of errors as a specific interface that has a form like the following: 
```
type VcdError struct {
    status int  // HTTP status code (could also be enum)
    reason string // HTTP reason (e.g., "Bad Request")
    msg string // full description of error
}

func (e *VcdError) Error() string { return e.msg }
```

See https://blog.golang.org/error-handling-and-go for more.  These conventions will lead to client code that is more flexible and more likely to be correct.

**Comment by dataclouder** [comment]
@hodgesrm,
Thanks for addressing the issue.
I fully agree on the assessment and the purpose of this issue.
Shall we convert it into an action item?

**Comment by vbauzysvmware** [comment]
@dataclouder I think we can close this as we have another issue to tackle this

**Comment by dataclouder** [comment]
This issue was already addressed by the refactoring of the `Get*` methods
